posthog/models/async_deletion/delete.py

```diff
@@ -1,146 +1,63 @@
-        with_data = set((team_id, str(key)) for team_id, key in _verify_by_column(columns, async_deletions))
-        return [row for row in async_deletions if (row.team_id, row.key) not in with_data]
-    else:
-        return []
-
-
-def _verify_by_column(distinct_columns: str, async_deletions: List[AsyncDeletion]) -> Set[Tuple[Any, ...]]:
-    conditions, args = _conditions(async_deletions)
-    clickhouse_result = sync_execute(
-        f"""
-        SELECT DISTINCT {distinct_columns}
-        FROM events
-        WHERE {" OR ".join(conditions)}
-        """,
-        args,
-    )
-    return set(tuple(row) for row in clickhouse_result)
-
-
-def _column_name(async_deletion: AsyncDeletion):
-    assert async_deletion.deletion_type in (DeletionType.Person, DeletionType.Group)
-    if async_deletion.deletion_type == DeletionType.Person:
-        return "person_id"
-    else:
-        return f"$group_{async_deletion.group_type_index}"
-
-
-def _conditions(async_deletions: List[AsyncDeletion]) -> Tuple[List[str], Dict]:
-    conditions, args = [], {}
-    for i, row in enumerate(async_deletions):
-        condition, arg = _condition(row, str(i))
-        conditions.append(condition)
-        args.update(arg)
-    return conditions, args
-
-
-def _condition(async_deletion: AsyncDeletion, suffix: str) -> Tuple[str, Dict]:
-    if async_deletion.deletion_type == DeletionType.Team:
-        return f"team_id = %(team_id{suffix})s", {f"team_id{suffix}": async_deletion.team_id}
-    else:
-        return (
-            f"(team_id = %(team_id{suffix})s AND {_column_name(async_deletion)} = %(key{suffix})s)",
-            {f"team_id{suffix}": async_deletion.team_id, f"key{suffix}": async_deletion.key},
-        )
+class AsyncDeletionProcess(ABC):
+    def __init__(self) -> None:
+        super().__init__()
+
+    def run(self):
+        queued_deletions = list(AsyncDeletion.objects.filter(delete_verified_at__isnull=True))
+        self.process(queued_deletions)
+
+    def mark_deletions_done(self):
+        """
+        Checks and updates `delete_verified_at` for deletions
+        """
+        to_verify = []
+        unverified = self._fetch_unverified_deletions_grouped()
+
+        for (deletion_type, _), async_deletions in unverified.items():
+            to_verify.extend(self._verify_by_group(deletion_type, async_deletions))
+
+        if len(to_verify) > 0:
+            AsyncDeletion.objects.filter(pk__in=[row.pk for row in to_verify]).update(delete_verified_at=timezone.now())
+            logger.info(
+                "Updated `delete_verified_at` for AsyncDeletion",
+                {"count": len(to_verify), "team_ids": list(set(row.team_id for row in to_verify))},
+            )
+
+    def _fetch_unverified_deletions_grouped(self):
+        result = defaultdict(list)
+        for item in AsyncDeletion.objects.filter(delete_verified_at__isnull=True):
+            key = (item.deletion_type, item.group_type_index)
```

Comment:
I'm confused how the `group_type_index` comes into play here in the base class? How does this work for cohorts?

Reply:
group_type_index irrelelvant. Only used for group type deletions. However, it doesn't event seem like there are any group type async deletions. I'll leave it in so it doesn't disrupt anything unaccounted for but it should be fine. Cohort deletion check is just based on item.deletion_type